1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+
+For example:
+
+```powershell
+cmd1 || throw "cmd1 failed"
+```
+
+```powershell
+function Invoke-Command
+{
+    cmd1 && return
+
+    cmd2
+}
+```
+
+```powershell
+foreach ($v in 1..100)
+{
+    cmd1 $v && break
+}
+```
+
+These would only be allowed *at the end* of chains because:
+
+- Control flow makes any invocation after the statement unreachable
+- `return` and `throw` allow pipelines as subordinate expressions,
+   meaning it would be grammatically impossible to use those statements mid-chain.
+   See below for more details.
+
+#### Reasons against
+
+- `throw` and `return` currently allow a subordinate pipeline:
+
+   ```powershell
+   throw "Bad"
+   ```
+
+   ```powershell
+   throw 1,2,3 | Write-Output
+   ```
+
+   ```powershell
+   throw "Useless" & # Throws a background job...
+   ```
+
+   If we keep the principle that anywhere a pipeline is allowed,
+   a pipeline chain is now allowed, then constructions like this are possible:
+
+   ```powershell
+   cmd1 || throw "Bad" && cmd3
+   ```
+
+   However the grouping is this:
+
+   ```none
+   [cmd1 || throw ["Bad" && cmd3]]
+   ```
+
+   This could lead to confusing semantic corner cases
+
+- The compromise nature of this approach means the PowerShell Language
+  becomes more complicated and arguably less consistent,
+  both conceptually and in terms of maintenance.
+
+### Different evaluations of command "success"
+
+The current proposal is to simply use `$?` to determine chain continuation.
+
+Alternatives include:
+
+- `$LASTEXITCODE`
+- Whether errors have been written
+- A new custom semantics
+
+A problem is that exit code semantics assume
+that a non-zero exit code means command failure.
+POSIX shells also make this assumption,
+but the convention may not be as widespread on Windows platforms.
+
+#### Reasons against
+
+- `$LASTEXITCODE` is specific to native commands
+  and use with cmdlets and functions may lead to unexpected results.
+  For example:
+
+  ```powershell
+  cmd_that_fails
+  Write-Output "SUCCESS" && Write-Output "ALSO SUCCESS"
+  ```
+
+  Will never write `"ALSO SUCCESS"` if `$LASTEXITCODE` is used.
+
+- Commands can fail without writing an error
+  and some utilities will write to standard error without having failed
+  (`time` being a good example on \*nix).
+
+- Having chain operators use their own failure semantics
+  would create more conceptual complexity
+  and be inconsistent with the established `$?` concept.
+
+- Allowing configurability of the success determination would
+  likely only make sense with a per-command configuration.
+  This would already be achieved simply with the proposed `$?`
+  semantics using a wrapper function.
+
+### Terminology
+
+We should have a unified terminology
+to describe these operators in PowerShell
+for use with:
+
+- The experimental feature name
+- The about_ topic
+- Web searchability
+
+This RFC proposes **Pipeline chain operators**.
+Other possibilities are:
+
+- AND-OR lists
+- Command chain operators

```

Comment:
I'd vote for this, with short-circuit operators being a close second for me, I think.